src/py/flwr_tool/update_changelog.py

```diff
@@ -0,0 +1,225 @@
+    token = next(
+        (token for token, marker in token_markers.items() if marker in entry_text), None
+    )
+
+    return entry_text, token
+
+
+def _update_changelog(prs):
+    """Update the changelog file with entries from provided pull requests."""
+    with open(CHANGELOG_FILE, "r+", encoding="utf-8") as file:
+        content = file.read()
+        unreleased_index = content.find("## Unreleased")
+
+        if unreleased_index == -1:
+            print("Unreleased header not found in the changelog.")
+            return
+
+        # Find the end of the Unreleased section
+        next_header_index = content.find("##", unreleased_index + 1)
+        next_header_index = (
+            next_header_index if next_header_index != -1 else len(content)
+        )
+
+        for pr_info in prs:
+            pr_entry_text, category = _extract_changelog_entry(pr_info)
+
+            # Skip if PR should be skipped or already in changelog
+            if category == "skip" or f"#{pr_info.number}]" in content:
+                continue
+
+            pr_reference = _format_pr_reference(
+                pr_info.title, pr_info.number, pr_info.html_url
+            )
+
+            # Process based on category
+            if category in ["general", "baselines", "examples", "sdk", "simulations"]:
+                entry_title = _get_category_title(category)
+                content = _update_entry(
+                    content,
+                    entry_title,
+                    pr_info,
+                    unreleased_index,
+                    next_header_index,
+                )
+
+            elif pr_entry_text:
+                content = _insert_new_entry(
+                    content, pr_info, pr_reference, pr_entry_text, unreleased_index
+                )
+
+            else:
+                content = _insert_entry_no_desc(content, pr_reference, unreleased_index)
+
+            next_header_index = content.find("##", unreleased_index + 1)
+            next_header_index = (
+                next_header_index if next_header_index != -1 else len(content)
+            )
+
+        # Finalize content update
+        file.seek(0)
+        file.write(content)
+        file.truncate()
+
+    print("Changelog updated.")
+
+
+def _get_category_title(category):
+    """Get the title of a changelog section based on its category."""
+    headers = {
+        "general": "General improvements",
+        "baselines": "General updates to Flower Baselines",
+        "examples": "General updates to Flower Examples",
+        "sdk": "General updates to Flower SDKs",
+        "simulations": "General updates to Flower Simulations",
+    }
+    return headers.get(category, "")
+
+
+def _update_entry(
+    content, category_title, pr_info, unreleased_index, next_header_index
+):
+    """Update a specific section in the changelog content."""
+    section_index = content.find(category_title, unreleased_index, next_header_index)
+    if section_index != -1:
```

Comment:
just want to mention a small tick i have discovered recently. the two lines can be combined in one by using the walrus operator `:=` that has been introduced in python 3.8.

(i haven't tested this line)
```suggestion
    if (section_index := content.find(category_title, unreleased_index, next_header_index)) != -1:
```
but i have no strong preference using it here so we can also leave the code as is.

Reply:
Oh nice! I'll test it out, thanks!